rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+---
+feature: no-direct-pushes
+start-date: 2023-07-21
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Require pull requests for all Nixpkgs commits.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently 197 people with commit access to Nixpkgs, all of whom can push directly to any branch without a pull request.
+Such pushes generally do not notify anybody and do not trigger CI.
+
+This makes such commits susceptible to:
+- Be anonymous
+- Include malicious code
+- Be broken
+- Have poor code quality
+
+While requiring pull requests isn't a panacea, it does help by:
+- Notifying [code owners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)
+- Running CI (though there's no requirement for it to succeed, see [future work][future])
+- Tying commits to a GitHub account
+- Being discoverable in the pull request list
```

Comment:
Looking through the entire pull request list doesn't seem to be much more practical than the commit list (which is also searchable, after all, and PR search is also buggy anyway)

Reply:
The pull request list is definitely a lot more discoverable than the commit list: Pull requests have various metadata (authors, labels, update time, etc.) you can filter and sort by. In addition, people have the habit of checking the pull request list, not the commit list. I don't think this should be controversial

Reply:
Labels being applicable is a great point — maybe make it explicit?

Reply:
Something about making contributions follow a common harmonized workflow? E.g. labels, reviews, etc. (Feel free to phrase in a good way.)

Reply:
I know of several people who have the pull requests page open a lot of the time. That's a lot of eyes on new changes, and I don't think anybody monitors commits in a similar way, especially since those easily get drowned in merge commits from other PRs

Reply:
I now expanded on this in https://github.com/NixOS/rfcs/pull/156/commits/becc35e5b64712c15df0fc65c3402f17bf67f351